diff --git a/src/sse/exec.ml b/src/sse/exec.ml
--- a/src/sse/exec.ml
+++ b/src/sse/exec.ml
@@ -656,7 +656,7 @@
               Path.set_pc path addr;
               (Interpreter.dispatch [@tailcall]) path fiber (module D)
           | Item { elt = Some code; _ } ->
-              let fiber = Disassembly.disassemble_from code addr in
+              let fiber = Disassembly.fetch_no_link code addr in
               Virtual_address.Htbl.add env.rocache addr fiber;
               Path.set_pc path addr;
               (Interpreter.dispatch [@tailcall]) path fiber (module D)
